dvc/utils/fs.py

```diff
@@ -15,23 +16,33 @@ def get_inode(path):
     return inode
 
 
-def get_mtime_and_size(path):
+def get_mtime_and_size(path, search_for_ignores=True):
     size = os.path.getsize(path)
-    mtime = os.path.getmtime(path)
 
     if os.path.isdir(path):
-        for root, dirs, files in os.walk(str(path)):
-            for name in dirs + files:
-                entry = os.path.join(root, name)
+        files_mtimes = {}
+        for root, dirs, files in dvc_walk(
+            str(path), search_for_ignores=search_for_ignores
+        ):
+            for dir in dirs:
+                entry = os.path.join(root, dir)
+                size += os.path.getsize(entry)
+
+            for file in files:
+                entry = os.path.join(root, file)
                 stat = os.stat(entry)
                 size += stat.st_size
-                entry_mtime = stat.st_mtime
-                if entry_mtime > mtime:
-                    mtime = entry_mtime
+                files_mtimes[entry] = stat.st_mtime
+        # we need to take into account that analysed dir can have .dvcignored
```

Comment:
Let's add a better explanation why we do this. Also, `files_mtimes` dict is not sorted, you need to use a list and actually sort it by relpath. See https://github.com/iterative/dvc/blob/master/dvc/remote/local.py#L251 . Otherwise the result of dict_md5 might not be reproducible. Let's add a test that is making sure that mtime_and_size is the same for the same dir.

Reply:
@pared pointed out that I'm wrong about this. Indeed, dict representation _is_ sorted in dict_md5 itself, so we should be fine. The reason I've used lists and sorted them in dir cache is that there I have opted for storing {"relpath": relpath, "md5": md5} dicts just in case we would need to add something else to them later. It has been a long time and we didn't add anything there yet, so that might've been for nothing :slightly_smiling_face: